### ISSUE TITLE
Spill to disk is inefficient for small byte-values

### ISSUE THREAD

**Issue by mrocklin** [dask/distributed#525]
As of #485 workers can spill excess data to disk.  They currently evict tasks from a fixed pool of memory to disk with a simple LRU policy.  Task results are stored on disk as single files.  This is currently done using the composable MutableMappings in a tiny project, [zict](https://github.com/mrocklin/zict).

```python
if memory_limit:
    from zict import Func, Buffer, File
    path = os.path.join(self.local_dir, 'storage')
    memory = dict()
    disk = File(path)
    disk = Func(dumps_to_disk, loads_from_disk, disk)
    self.data = Buffer(memory, disk, int(float(memory_limit)), weight)  # LRU from memory to disk
else:
    self.data = {}
```

This policy is decent for large results (larger than say, a megabyte) but will slow down when trying to save many small results due to file system overhead.  This case specifically occurs when performing shuffles, such as with dask.array rechunkings, or dask.bag/dataframe joins, where we have possibly millions of very small results.  Saving thousands of tiny files is *very* slow on most file systems.  [See blogpost](http://matthewrocklin.com/blog/work/2015/12/29/disk-bandwidth).

Single-machine key-value store databases provide decently efficient random write and read access to small values on disk.  There are *many* solutions here.  I'm somewhat partial to embedded databases like LevelDB (which now has many more modern competitors) that run inside the same process because they don't require significant setup, run in a well contained way, and are easy to use.

So one solution here would be to optionally replace the `File(MutableMapping)` from zict that uses the filesystem with a MutableMapping that was backed by some more efficient embedded database.  This would require learning a bit about different key-value stores and choosing a good one based on the many-small writes/reads workloads that dask.distributed might push onto it, finding a way to combine it with the LRU scheme (or some other scheme) used currently (this is easy if you can make it follow a `MutableMapping` interface), and then hooking up the necessary keywords to allow users to access this option easily.

**Comment by mrocklin** [comment]
An alternative to this would be to build a specialized shuffle into dask.distributed.  If we want to make dask.dataframe joins as fast as what you find in a database then something like this probably necessary.  The drawback is that the specialized shuffle would be entirely separate from the task scheduling system.  This is how databases/Spark do it.  Historically Dask has always been "just a task scheduler" eases maintenance and keeps all of our optimizations (like the optimization above) useful in other situations than just database cases.  

However there is always a choice.  Dask could pivot and jump into the database game.  I'm personally not excited about this (there are many excellent databases and few excellent task schedulers) but other people might feel differently.  Building a shuffle is still an option.  It might even be doable without much pain.

**Comment by mrocklin** [comment]
I've heard good things about rocksdb from a couple of sources now.  

There appears to be an ubuntu package and a decent python library

**Comment by kszucs** [comment]
I think dask would gain a lot of popularity as a "drop-in-replacement" for pyspark dataframes. A lot of people uses pyspark for biggish problems just because there are no alternatives in the python ecosystem. They need to deal a lot of overhead with learning/experimenting/digging/configuring pyspark, jvm, OOMs etc (it's a nightmare - especially after using pandas). I want no more jars, class names taking miles to write, hundreds of configuration options. I just want to use pandas interface on distributed data, efficiently. Dask distributed is really close, in fact I'd threw out all of our problematic pyspark jobs if there were a fast enough shuffle mechanism in distributed.

> Historically Dask has always been "just a task scheduler" eases maintenance and keeps all of our optimizations (like the optimization above) useful in other situations than just database cases.

Historically Spark was a proof-of-concept framework for mesos :)

I can see _dask_ as the de facto standard distributed solution for python, since it already has bags, arrays, dataframes. I think these existing containers implies that dask is not _just_ a task scheduler.

On the other hand, doesn't disk shuffle on disk (via partd) violate this policy? 

I'd really love to see both, a more efficient spill-to-disk and a more efficient distributed shuffle in the near future.

**Comment by mrocklin** [comment]
Yeah, to be clear I'd like to restrict this issue is about the specific case of spill-to-disk-workers.  I'd like to focus the spark conversation elsewhere.  I apologize for adding the shuffle note above.

If you want to start the shuffle work please be my guest.  I agree that it would have high value.  It also has high development and maintenance cost.  It is nowhere on my personal roadmap for the project.  Would love to have active contributors work on this independently.

**Comment by pitrou** [comment]
There's a simple benchmark of key/value stores written in Python there: http://charlesleifer.com/blog/completely-un-scientific-benchmarks-of-some-embedded-databases-with-python/

I'll try to reproduce, and see whether the benchmark makes sense. It's interesting that the venerable `gdbm` appears the fastest of the bunch, at least in this specific test.

**Comment by mrocklin** [comment]
I spent some plane ride a while ago starting on a RocksDB MutableMapping here: https://github.com/mrocklin/zict/pull/2

Not particularly mature but might be useful for a start

**Comment by pitrou** [comment]
Another contestant: https://lsm-db.readthedocs.io/en/latest/index.html
Another source of information about embedded databases: http://engine.so/

**Comment by jcrist** [comment]
I've also heard good things about lmdb. Python bindings: https://lmdb.readthedocs.io/en/release/. From reading the docs it may not be appropriate for this use case though. It seems to be optimized for reading over writing, and small values over large. Benchmark: http://104.237.133.194/bench/ondisk/.

**Comment by pitrou** [comment]
Ok, here is a benchmark script of value key-value stores: https://gist.github.com/pitrou/9c9734dc266e3dcb2c6383102245045c

It uses fixed-length keys and variable-length values. The length of values is distributed according to a lognormal distribution. You can of course tune the parameters in the script. Both sequential (sorted) and random access is benchmarked. The write workload only writes new values, not existing ones.

Here are the results on my machine (64-bit Linux, SSD): https://gist.github.com/pitrou/beb124b8626fa390d5d26a5b0f6d8d5d

All tested stores look convincing: a couple µs per operation means the overhead is small compared to regular Python bytecode execution. Kyoto Cabinet is practically unusable for us as it's GPL-licensed. LMDB appears the best of the bunch on this particular benchmark. RocksDB is surprisingly a bit slower than LevelDB.

It would be nice to know if others get different results. You may have to `pip install` some libraries...

**Comment by pitrou** [comment]
I got similar results for LMDB on a 64-bit Windows VM. The other libraries I wasn't able to install on Windows...

**Comment by pitrou** [comment]
A quick test shows that LMDB is also very fast with large (multi-MB) values, actually as fast as directly writing files, achieving more than 1 GB/s.

**Comment by pitrou** [comment]
Note the LMDB figures were obtained with `writemap=True`, which is apparently unadvised on some systems ("This option may cause filesystems that don’t support sparse files, such as OSX, to immediately preallocate map_size= bytes of underlying storage").

**Comment by mrocklin** [comment]
Looks like LMDB is already on conda-forge:
https://github.com/conda-forge/lmdb-feedstock

**Comment by pitrou** [comment]
This is the LMDB shared library, not the Python package.

**Comment by mrocklin** [comment]
Hrm, that comment was actually written and sent several hours ago, before you authored the python-lzma staged recipe.  Wasn't intended to dissuade you from https://github.com/conda-forge/staged-recipes/pull/1809

**Comment by mrocklin** [comment]
Do these change significantly when the byte values are in the 100MB range?
This range is as important as the the small-byte range.

**Comment by mrocklin** [comment]
For the GB/s write speeds, does your computer have NVMe?  If not then we're
probably seeing write-to-ram speeds rather than write-to-disk.  Are you
flushing at the end of the test?

**Comment by pitrou** [comment]
No, this is on a SATA SSD. I am not flushing at the end of the test: is it required for this workload?

**Comment by mrocklin** [comment]
It's not required, but it may change results.  Writing to memory might have different performance characteristics than writing to disk.  Flushing would give some indication if there was a significant difference.

**Comment by pitrou** [comment]
Ideally we would test under limited memory conditions, to let the database (or the OS) use its own flushing heuristics. Unfortunately I don't know how to limit the amount of physical RAM available to the benchmark process (`ulimit -m` doesn't seem to work), and a multi-GB test would be quite slow...

**Comment by pitrou** [comment]
That said, we can simply err on the side of safety and use the file backend for large data values (> 1MB perhaps).

**Comment by mrocklin** [comment]
There are some times when we need to write many small values.  This happens during operations like row-major to column-major rechunkings in dask.array or joins and groupbys in dask.dataframe.

I don't think we need to spend too much more time benchmarking though.  Anything will be a large improvement and it seems to be fairly easy to change this in the future.
